sw/device/silicon_creator/rom_ext/rescue.c

```diff
@@ -0,0 +1,228 @@
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/silicon_creator/rom_ext/rescue.h"
+
+#include "sw/device/lib/base/memory.h"
+#include "sw/device/silicon_creator/lib/dbg_print.h"
+#include "sw/device/silicon_creator/lib/drivers/flash_ctrl.h"
+#include "sw/device/silicon_creator/lib/drivers/retention_sram.h"
+#include "sw/device/silicon_creator/lib/drivers/rstmgr.h"
+#include "sw/device/silicon_creator/lib/xmodem.h"
+
+#include "flash_ctrl_regs.h"
+
+// All of the xmodem functions accept an opaque iohandle pointer.
+// The iohandle is used to facilitate unit tests and doesn't have
+// any function in real firmware.
+#define iohandle NULL
+
+static rescue_state_t rescue_state;
+
+rom_error_t flash_firmware_block(rescue_state_t *state) {
+  if (state->flash_offset == 0) {
+    flash_ctrl_data_default_perms_set((flash_ctrl_perms_t){
+        .read = kMultiBitBool4True,
+        .write = kMultiBitBool4True,
+        .erase = kMultiBitBool4True,
+    });
+    for (uint32_t addr = state->flash_start; addr < state->flash_limit;
+         addr += FLASH_CTRL_PARAM_BYTES_PER_PAGE) {
+      HARDENED_RETURN_IF_ERROR(
+          flash_ctrl_data_erase(addr, kFlashCtrlEraseTypePage));
+    }
+    state->flash_offset = state->flash_start;
+  }
+  if (state->flash_offset < state->flash_limit) {
+    HARDENED_RETURN_IF_ERROR(flash_ctrl_data_write(
+        state->flash_offset, sizeof(state->data) / sizeof(uint32_t),
+        state->data));
+    state->flash_offset += sizeof(state->data);
+  } else {
+    return kErrorRescueImageTooBig;
+  }
+  return kErrorOk;
+}
+
+rom_error_t flash_owner_block(rescue_state_t *state) {
+  // FIXME: validate that we're in a state capable of accepting an owner
+  // block and validate the block before flashing it.
+  HARDENED_RETURN_IF_ERROR(flash_ctrl_info_erase(&kFlashCtrlInfoPageOwnerSlot1,
+                                                 kFlashCtrlEraseTypePage));
+  HARDENED_RETURN_IF_ERROR(flash_ctrl_info_write(
+      &kFlashCtrlInfoPageOwnerSlot1, 0, sizeof(state->data) / sizeof(uint32_t),
+      state->data));
+  return kErrorOk;
+}
+
+static void validate_mode(uint32_t mode, rescue_state_t *state) {
+  char *m = (char *)&mode;
+  dbg_printf("\r\nmode: %C%C%C%C\r\n", m[3], m[2], m[1], m[0]);
+  switch (mode) {
+    case kRescueModeBootLog:
+      dbg_printf("ok: receive boot_log via xmodem-crc\r\n");
+      break;
+    case kRescueModeBootSvcRsp:
+      dbg_printf("ok: receive boot_svc response via xmodem-crc\r\n");
+      break;
+    case kRescueModeBootSvcReq:
+      dbg_printf("ok: send boot_svc request via xmodem-crc\r\n");
+      break;
+    case kRescueModeOwnerBlock:
+      dbg_printf("ok: send owner_block via xmodem-crc\r\n");
+      break;
+    case kRescueModeFirmware:
+      dbg_printf("ok: send firmware via xmodem-crc\r\n");
+      break;
+    case kRescueModeReboot:
+      dbg_printf("ok: reboot\r\n");
+      break;
+    case kRescueModeDWIM:
+      // Easter egg :)
```

Comment:
Any insights for the people who don't know what this refers to? :)

Reply:
DWIM is a hypothetical CPU opcode: Do What I Mean.  In some architectures, it is named "RPI" (Read Programmer Intent).